system-addon/lib/TopStoriesFeed.jsm

```diff
@@ -277,6 +297,83 @@ this.TopStoriesFeed = class TopStoriesFeed {
     this.newTabsSinceSpoc++;
   }
 
+  // Frequency caps are based on campaigns, which may include multiple spocs.
+  // We currently support two types of frequency caps:
+  // - lifetime: Indicates how many times spocs from a campaign can be shown in total
+  // - period: Indicates how many times spocs from a campaign can be shown within a period
+  //
+  // So, for example, the feed configuration below defines that for campaign 1 no more
+  // than 5 spocs can be show in total, and no more than 2 per hour.
+  // "campaign_id": 1,
+  // "caps": {
+  //  "lifetime": 5,
+  //  "campaign": {
+  //    "count": 2,
+  //    "period": 3600
+  //  }
+  // }
+  isBelowFrequencyCap(impressions, spoc) {
+    const campaignImpressions = impressions[spoc.spoc_meta.campaign_id];
+    if (!campaignImpressions) {
+      return true;
+    }
+
+    const lifeTimeCap = spoc.spoc_meta.caps && spoc.spoc_meta.caps.lifetime;
+    const campaignCap = (spoc.spoc_meta.caps && spoc.spoc_meta.caps.campaign) || {};
+
+    const lifeTimeCapExceeded = () => campaignImpressions.length >= lifeTimeCap;
```

Comment:
Any reason of creating two functions here, looks like it just needs two booleans.

Reply:
Yes, just to get lazy evaluation. This way we don't have to calculate campaignCapExceeded if lifeTimeCapExceeded is already true.

Reply:
Of course, I could inline it instead of storing separately but found this to be easier to read.

Reply:
Let's just inline them for better readability.

Reply:
OK, it was an unnecessary optimization anyway. Fixing it up.